### PR TITLE
Shopify Client

### PR DESCRIPTION
**Pull request by sbu-WBT**
Shopify needs the client_id and client_secret inside the body see https://shopify.dev/tutorials/authenticate-with-oauth#step-3-confirm-installation

This is also rfc compliant https://tools.ietf.org/html/rfc6749#section-2.3.1

### CODE REVIEW

**Comment by sbu-WBT** [comment]
Is the failure my fault? I have to say that I can't follow the log errors

**Comment by okrad** [comment]
Hi @sbu-WBT, thank you for the PR!

I like the idea, I just have a couple observations:

1. Instead of using a boolean parameter, I think it could be better to use an enumerated value, that currently can take only 2 values (conveying the _"pass credentials in the header"_ and _"pass credentials in the body"_ meanings), but in the future could be extended to support other values (for example _"pass credentials in the querystring"_ - I know it sounds silly but there are many non-standard servers out there...). In this case I think a better name for the parameter could be ``credentialsLocation`` or something like that. The values could be codified as constants in the OAuth2Client class (``CREDENTIALS_LOCATION_HEADER`` and ``CREDENTIALS_LOCATION_BODY``?)

2. IMHO that parameter could be a property of the Auth2Client class, meaning that we could pass it directly in the client constructor, so that it could be seen as a "configuration" of the client. Something like:

```dart
OAuth2Client(
      {@required this.authorizeUrl,
      @required this.tokenUrl,
      this.refreshUrl,
      this.revokeUrl,
      this.credentialsLocation, //Sets the location of the credentials for the whole client
      @required this.redirectUri,
      @required this.customUriScheme}) {
...
}
```
This way we don't have to pass the parameter to the OAuth2Helper and from there on to all the client methods.
Your Shopify client could then be changed to:

```dart
import 'package:meta/meta.dart';

import 'oauth2_client.dart';

class ShopifyOAuth2Client extends OAuth2Client {
  ShopifyOAuth2Client({
    @required String shop,
    @required String redirectUri,
    @required String customUriScheme,
  }) : super(
          authorizeUrl: 'https://$shop.myshopify.com/admin/oauth/authorize',
          tokenUrl: 'https://$shop.myshopify.com/admin/oauth/access_token',
          redirectUri: redirectUri,
          credentialsLocation: OAuth2Client.CREDENTIALS_LOCATION_BODY,
          customUriScheme: customUriScheme,
        );
}

```

What do you think? If it sounds good to you, are you willing to update your code and update the PR?

**Comment by sbu-WBT** [comment]
I didn't thought on other possible ways to authenticate because only these are mentioned in the rfc. Bu if you want, I can transform it to an enum.

I thought the same of putting it inside the client. Then I saw the `enablePKCE` option in the helper and thought that these are kind of the same type of configuration. I will move it to the client

**Comment by okrad** [comment]
Fine, thank you!

File: lib/oauth2_client.dart [hunk 1]
```diff
@@ -41,14 +47,18 @@ class OAuth2Client {
...
 
   OAuth2Client(
       {@required this.authorizeUrl,
       @required this.tokenUrl,
       this.refreshUrl,
       this.revokeUrl,
       @required this.redirectUri,
-      @required this.customUriScheme}) {
+      @required this.customUriScheme,
+      this.credentialsLocation,
```
Comment (okrad) [inline lib/oauth2_client.dart:56]: ... otherwise, if the ``credentialsLocation`` param isn't passed to the constructor, the class property will be overridden to null.
```suggestion
      this.credentialsLocation = CredentialsLocation.HEADER,
```

File: lib/oauth2_client.dart [hunk 2]
```diff
@@ -41,14 +44,17 @@ class OAuth2Client {
   Map<String, String> _accessTokenRequestHeaders;
 
   WebAuth webAuthClient;
+  CredentialsLocation credentialsLocation = CredentialsLocation.HEADER;
```
Comment (okrad) [inline lib/oauth2_client.dart:47]: ``credentialsLocation`` should be initialized in the constructor (see below)...

**Review by okrad — CHANGES_REQUESTED** [review]
Thank you, there's an issue with the ``credentialsLocation`` initialization, that's why the tests are failing... Can you fix it?
Moreover, can you provide at least a test case?

**Comment by sbu-WBT** [comment]
I tried adding a test but failed😅 I have no clue why `NoSuchMethodError: The getter 'statusCode' was called on null.` gets thrown. I just copied an existing test

**Comment by okrad** [comment]
Thank you @sbu-WBT!
No problem for the tests, I added a couple of methods myself...
